--- a/openvdb/openvdb/python/CMakeLists.txt
+++ b/openvdb/openvdb/python/CMakeLists.txt
@@ -93,6 +93,16 @@ else()
   install(TARGETS openvdb_python DESTINATION ${VDB_PYTHON_INSTALL_DIRECTORY})
 endif()
 
+nanobind_add_stub(
+  openvdb_stub
+  INSTALL_TIME
+  MODULE openvdb
+  OUTPUT_PATH openvdb
+  PYTHON_PATH "."
+  MARKER_FILE py.typed
+  RECURSIVE
+)
+
 # pytest
 if(OPENVDB_BUILD_PYTHON_UNITTESTS)
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -30,15 +30,10 @@ slack = "https://slack.aswf.io/"
 wheel.packages = []
 
 [tool.scikit-build.cmake.define]
-OPENVDB_CORE_STATIC="OFF"
+OPENVDB_CORE_SHARED="OFF"
 USE_EXPLICIT_INSTANTIATION="OFF"
 DISABLE_DEPENDENCY_VERSION_CHECKS="ON"
 OPENVDB_USE_DELAYED_LOADING="OFF"
 OPENVDB_BUILD_PYTHON_MODULE="ON"
 USE_NUMPY="ON"
 
-USE_NANOVDB="ON"
-NANOVDB_USE_CUDA="ON"
-NANOVDB_BUILD_PYTHON_MODULE="ON"
-NANOVDB_USE_OPENVDB="ON"
-
-